test/services/send_transaction_file.service.test.js

```diff
@@ -0,0 +1,112 @@
+'use strict'
+
+// Test framework dependencies
+const Lab = require('@hapi/lab')
+const Code = require('@hapi/code')
+const Sinon = require('sinon')
+
+const { describe, it, beforeEach, afterEach } = exports.lab = Lab.script()
+const { expect } = Code
+
+// Test helpers
+const {
+  BillRunHelper,
+  DatabaseHelper,
+  GeneralHelper,
+  RegimeHelper
+} = require('../support/helpers')
+
+// Things we need to stub
+const { GenerateTransactionFileService, SendFileToS3Service } = require('../../app/services')
+
+// Thing under test
+const { SendTransactionFileService } = require('../../app/services')
+
+describe('Send Transaction File service', () => {
+  let regime
+  let billRun
+  let generateStub
+  let sendStub
+
+  beforeEach(async () => {
+    await DatabaseHelper.clean()
+
+    regime = await RegimeHelper.addRegime('wrls', 'WRLS')
+    billRun = await BillRunHelper.addBillRun(regime.id, GeneralHelper.uuid4())
+
+    generateStub = Sinon.stub(GenerateTransactionFileService, 'go').returns('stubFilename')
+    sendStub = Sinon.stub(SendFileToS3Service, 'go').returns(true)
+  })
+
+  afterEach(() => {
+    Sinon.restore()
+  })
+
+  describe('When a valid bill run is specified', () => {
+    beforeEach(async () => {
+      billRun.status = 'pending'
+    })
+
+    describe('and a transaction file is required', () => {
+      beforeEach(async () => {
+        billRun.fileReference = 'FILE_REFERENCE'
+      })
+
+      it('generates a transaction file', async () => {
+        await SendTransactionFileService.go(regime, billRun)
+
+        expect(generateStub.calledOnce).to.be.true()
+        expect(generateStub.getCall(0).firstArg).to.equal(`${billRun.fileReference}.dat`)
+      })
+
+      it('sends the transaction file', async () => {
+        await SendTransactionFileService.go(regime, billRun)
+
+        expect(sendStub.calledOnce).to.be.true()
+        expect(sendStub.getCall(0).firstArg).to.equal('stubFilename')
+        expect(sendStub.getCall(0).args[1]).to.equal(`${regime.slug}/transaction/${billRun.fileReference}.dat`)
+      })
+    })
+
+    describe("and a transaction file isn't required", () => {
+      it("doesn't try to generate a transaction file", async () => {
+        await SendTransactionFileService.go(regime, billRun)
+
+        expect(generateStub.notCalled).to.be.true()
+      })
+
+      it("doesn't try to send the transaction file", async () => {
+        await SendTransactionFileService.go(regime, billRun)
+
+        expect(sendStub.notCalled).to.be.true()
+      })
+    })
+
+    it("sets the bill run status to 'billed'", async () => {
+      await SendTransactionFileService.go(regime, billRun)
+
+      // The service returns a response immediately to avoid the user waiting for a response, so we need to pause
```

Comment:
Think I need some help with this one.

The `/send` endpoint after awaiting `SendBillRunReferenceService` completing will then call `SendTransactionFileService` but not wait for it as it will be returning a response to the user.

In the meantime, `SendTransactionFileService` will have started its thing. I think that means there is no pressing need for it to return immediately once the file has been sent. I think we can allow it to `await` the status being updated before completing. Or have I misunderstood something?!

Reply:
No, you're bang on! I just didn't consider that even though we won't be calling it and waiting for it finish, that doesn't mean we can't change it so we _can_ wait for it to finish if we want. So `SendTransactionFileService` now `awaits` the status-setting which means we can just wait for it to finish and then check the status.